### PR TITLE
Support batched arguments in dynamic_slice/update_slice

### PR DESCRIPTION
**Pull request by jakevdp**
This makes `lax.dynamic_slice` and `lax.dynamic_update_slice` natively support straightforward batching in all arguments, where "straightforward" means all arguments have leading batch dimensions that match. For example:
```python
import jax
import jax.numpy as jnp

x = jnp.array([[1, 2, 3, 4, 5],
               [6, 7, 8, 9, 0]])
indices = [jnp.array([1, 2])]
out = jax.lax.dynamic_slice(x, indices, slice_sizes=(2,))
print(out)
# [[2 3]
#  [8 9]]

update = jnp.array([[10, 20],
                    [30, 40]])
out = jax.lax.dynamic_update_slice(x, update, indices)
print(out)
# [[ 1 10 20  4  5]
#  [ 6  7 30 40  0]]
```
Note also that in this means the primitives are closed under batching in such cases, so e.g. `vmap(dynamic_slice)` lowers to batched `dynamic_slice` rather than `gather`:
```python
from functools import partial
print(jax.make_jaxpr(jax.vmap(partial(jax.lax.dynamic_slice, slice_sizes=(2,))))(x, indices))
# { lambda ; a:i32[2,5] b:i32[2]. let
#     c:bool[2] = lt b 0
#     d:i32[2] = add b 5
#     e:i32[2] = select_n c b d
#     f:i32[2,2] = dynamic_slice[slice_sizes=(2,)] a e
#   in (f,) }
```
This new batch-awareness is not flexible enough to handle arbitrary batching, but this could be accomplished in the future by introducing new keywords (`operand_batch_dims` and `index_batch_dims`). We leave that discussion for a future PR.

Because the mlir primitives do not yet support batched inputs, the lowering rule still lowers to `gather`/`scatter` for these cases, but I've been in conversation with MLIR developers who are planning to support this style of batching natively in the future.

### CODE REVIEW

File: jax/_src/lax/slicing.py [hunk 1]
```diff
@@ -900,10 +927,28 @@ def _dynamic_slice_batching_rule(batched_args, batch_dims, *, slice_sizes):
...
+  assert bdim == 0
+  return result
+
+def _dynamic_slice_batching_rule(batched_args, batch_dims, *, slice_sizes):
+  # A dynamic slice is a special case of gather; we can delegate to the gather
+  # batching rule.
+  # TODO(phawkins): consider removing dynamic_slice entirely and using gather
+  # always.
+
+  # When all operands are batched, we can lower to a batcheed dynamic_slice.
```
Comment (hawkinsp) [inline jax/_src/lax/slicing.py:943]: sp. batched

**Review by hawkinsp — APPROVED** [review]
LGTM!

Obviously it would be even better if the operand and indices could be batched independently, but it's fine to leave that for the future. Add a TODO?

File: jax/_src/lax/slicing.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/lax/slicing.py]: Done and done. Thanks!

**Comment by jakevdp** [comment]
Thinking more about this... I don't think we should land this without the additional parameter to express general baching. The issue is that for sensible defaults of these parameters (i.e. set to empty tuple), the behavior for batched inputs as in this PR would have to change.

**Comment by jakevdp** [comment]
I'm going to close this because I'm not planning to work on this more any time soon.

**Pull request closed without merging**